<h1>Extracting Stock Data Using a Web Scraping</h1>


Not all stock data is available via API, so in this assignment we will use web-scraping to obtain financial data.\
Using beautiful soup we will extract historical share data from a web-page.

## Table of Contents

1. Downloading the Webpage Using Requests Library
2. Parsing Webpage HTML Using BeautifulSoup
3. Extracting Data and Building DataFrame

In [1]:
#!pip install pandas
#!pip install requests
!pip install bs4
#!pip install plotly

     |████████████████████████████████| 122kB 23.2MB/s eta 0:00:01
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/a0/b0/b2/4f80b9456b87abedbc0bf2d52235414c3467d8889be38dd472
Successfully built bs4


In [2]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

## Using Webscraping to Extract Stock Data Example


First we must use the `request` library to downlaod the webpage, and extract the text. We will extract Netflix stock data <https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/netflix_data_webpage.html>.


In [3]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/netflix_data_webpage.html"

data  = requests.get(url).text

Next we must parse the text into html using `beautiful_soup`


In [4]:
soup = BeautifulSoup(data, 'html5lib')

Now we can turn the html table into a pandas dataframe


In [5]:
netflix_data = pd.DataFrame(columns=["Date", "Open", "High", "Low", "Close", "Volume"])

# First we isolate the body of the table which contains all the information
# Then we loop through each row and find all the column values for each row
for row in soup.find("tbody").find_all('tr'):
    col = row.find_all("td")
    date = col[0].text
    Open = col[1].text
    high = col[2].text
    low = col[3].text
    close = col[4].text
    adj_close = col[5].text
    volume = col[6].text
    
    # Finally we append the data of each row to the table
    netflix_data = netflix_data.append({"Date":date, "Open":Open, "High":high, "Low":low, "Close":close, "Adj Close":adj_close, "Volume":volume}, ignore_index=True)    

We can now print out the dataframe


In [6]:
netflix_data.head()

Date    Open    High     Low   Close       Volume Adj Close
0  Jun 01, 2021  504.01  536.13  482.14  528.21   78,560,600    528.21
1  May 01, 2021  512.65  518.95  478.54  502.81   66,927,600    502.81
2  Apr 01, 2021  529.93  563.56  499.00  513.47  111,573,300    513.47
3  Mar 01, 2021  545.57  556.99  492.85  521.66   90,183,900    521.66
4  Feb 01, 2021  536.79  566.65  518.28  538.85   61,902,300    538.85

We can also use the pandas `read_html` function


In [7]:
read_html_pandas_data = pd.read_html(url)

Beacause there is only one table on the page, we just take the first table in the list returned


In [8]:
netflix_dataframe = read_html_pandas_data[0]

netflix_dataframe.head()

Date    Open    High     Low  Close* Adj Close**     Volume
0  Jun 01, 2021  504.01  536.13  482.14  528.21      528.21   78560600
1  May 01, 2021  512.65  518.95  478.54  502.81      502.81   66927600
2  Apr 01, 2021  529.93  563.56  499.00  513.47      513.47  111573300
3  Mar 01, 2021  545.57  556.99  492.85  521.66      521.66   90183900
4  Feb 01, 2021  536.79  566.65  518.28  538.85      538.85   61902300

## Using Webscraping to Extract Stock Data Exercise

We will use the `requests` library to download the webpage <https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/amazon_data_webpage.html>. Then, we'll save the text of the response as a variable named `html_data`.


In [17]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/amazon_data_webpage.html"

html_data  = requests.get(url).text

Parse the html data using `beautiful_soup`.


In [18]:
soup = BeautifulSoup(html_data, 'html5lib')

Let's see the content of the title attribute:


In [19]:
soup.title

<title>Amazon.com, Inc. (AMZN) Stock Historical Prices &amp; Data - Yahoo Finance</title>

Using beautiful soup extract the table with historical share prices and store it into a dataframe named `amazon_data`. The dataframe should have columns Date, Open, High, Low, Close, Adj Close, and Volume.


In [20]:
amazon_data = pd.DataFrame(columns=["Date", "Open", "High", "Low", "Close", "Volume"])

for row in soup.find("tbody").find_all("tr"):
    col = row.find_all("td")
    date = col[0].text #ADD_CODE
    Open = col[1].text #ADD_CODE
    high = col[2].text #ADD_CODE
    low = col[3].text #ADD_CODE
    close = col[4].text #ADD_CODE
    adj_close =col[5].text #ADD_CODE
    volume = col[6].text #ADD_CODE
    
    amazon_data = amazon_data.append({"Date":date, "Open":Open, "High":high, "Low":low, "Close":close, "Adj Close":adj_close, "Volume":volume}, ignore_index=True)

Print out a few rows of the `amazon_data` dataframe created.


In [21]:
amazon_data

Date      Open      High       Low     Close       Volume  \
0   Jan 01, 2021  3,270.00  3,363.89  3,086.00  3,206.20   71,528,900   
1   Dec 01, 2020  3,188.50  3,350.65  3,072.82  3,256.93   77,556,200   
2   Nov 01, 2020  3,061.74  3,366.80  2,950.12  3,168.04   90,810,500   
3   Oct 01, 2020  3,208.00  3,496.24  3,019.00  3,036.15  116,226,100   
4   Sep 01, 2020  3,489.58  3,552.25  2,871.00  3,148.73  115,899,300   
..           ...       ...       ...       ...       ...          ...   
56  May 01, 2016    663.92    724.23    656.00    722.79   90,614,500   
57  Apr 01, 2016    590.49    669.98    585.25    659.59   78,464,200   
58  Mar 01, 2016    556.29    603.24    538.58    593.64   94,009,500   
59  Feb 01, 2016    578.15    581.80    474.00    552.52  124,144,800   
60  Jan 01, 2016    656.29    657.72    547.18    587.00  130,200,900   

   Adj Close  
0   3,206.20  
1   3,256.93  
2   3,168.04  
3   3,036.15  
4   3,148.73  
..       ...  
56    722.79  
57    659.59  
58    593.64  
59    552.52  
60    587.00  

[61 rows x 7 columns]

Let's see what is the name of the columns of the dataframe


In [22]:
amazon_data.columns

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Adj Close'], dtype='object')

We can see that the `Open` of the last row of the amazon_data dataframe is 656.29.

__Great!__ This is the end of the notebook. In this code notebook some of knowledge required for a data scientist and some of the skills used by data scientists on a daily basis were shown and applied. The code and images were provided by IBM, and the development of the code and notebook, as well as some notes and editions, were carried out by me, [Saulo Villaseñor](https://www.linkedin.com/in/saulo-villase%C3%B1or-60669610a), so that this notebook is available to everybody and work as a reference for anyone who wishes to learn new skills.